### DISCUSSION TITLE
Equality checks and large data sets

### DISCUSSION THREAD

**Discussion by hypeJunction** [ai/nanostores#274] [General]
I am a working with a large dynamic deep nested map of data. I have read in some of the tickets that there is no batch update for maps and deep maps, because the expectation is that the UI framework should take care of throttling rerenders. That however isn't very efficient - I have thousands of updates within dozens of milliseconds and they freeze the app because of all the changes that happen to the map. Each change causes a chain reaction of notifying listeners, which then recompute and notify their listeners, and so it goes. For now, I have switched back to atom, replacing the entire state at once, which solves the freezing problem, but I suspect there is still too much unnecessary recomputation going on. 

If we look at the atom setter, we see this equality check:

```js
    set(data) {
      if ($atom.value !== data) {
        $atom.value = data
        $atom.notify()
      }
    },
```

Working with atoms, I can do some other equality checks from the outside. In case of computed values, however, it seems that there is no way to guarantee that values are identical even though values are equal.

I am wondering what your suggestion would be. I see 2 possible solutions:
1. Add support for batch updating maps before firing notify
2. Extend computed with an option to provide a custom function that would check equality before setting the atom value

**Comment by ai** [comment]
> Extend computed with an option to provide a custom function that would check equality before setting the atom value

[`computed`](https://github.com/nanostores/nanostores/blob/main/computed/index.js) is just syntax sugar on top of listener. You can create your own `computed` with any logic.

I think it is better than making API complex with more options.

> Add support for batch updating maps before firing notify

It is also nice idea, but we need research of API and how to implement it with less code.

**Comment by hypeJunction** [comment]
Tried to rework `computed`, but somehow the TypeScript magic got lost :( There is a considerable performance gain though. 
Will be off next week, so will revisit when I am back from holiday.

**Comment by hypeJunction** [comment]
After some more thinking, I have decided that `computed` is not necessarily where the bottleneck is. Ultimately, the problem is how the UI framework deals with updates. As we are using Vue.js, I have now written a custom adapter that also does the equality checks: https://github.com/nanostores/vue/issues/13
